Memfault: Add support for Wi-Fi and nRF7002 DK, move sample to `debug/`

jtguggedal opened the pull request:
The sample is moved from `nrf9160/` to `debug/` as an initial step
in making it work for mulltiple network technologies.

A few ifdefs are added to main.c, and project configuration
is split into common configs in `prj.con`f and board specific
configs in respective `boards/<target>.conf` files.

---

Use HW ID library for device ID.
The IMEI is still used for nRF91 boards, only now through
the HW ID library.

---

Add support for the nrf7002dk_nrf5350_cpuapp build target.

---

Network-related code is reworked to make it simpler to add new
network types and clean up main.c.
Currently LTE and Wi-Fi are supported, using the nRF9160 DK,
Thingy:91 and nRF7002 DK.

jtguggedal commented:
@KAGA164 Could you have a look and check if the changes to the integration code is okay also for Bluetooth?
With the rework of the connectivity code in the sample, it should actually also be possible to include Bluetooth in the same sample, but not sure it makes sense.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11019/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11019/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11019/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-11019
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -411,6 +411,11 @@ Other samples
     * A workaround for the hardware `Errata 257`_ of the nRF52811 chip.
     * A workaround for the hardware `Errata 117`_ of the nRF5340 chip.
 
+* :ref:`memfault_sample` sample:
```
Shouldn't this be under the 9160 samples, as that is where it used to be and mention there that the sample has been not only moved to another location within NCS, but also moved under Other samples in the documentation structure.

jtguggedal replied:
I was wondering the same myself. I'll come up with a suggestion to solve both the changelog and Debug samples subpage

peknis commented on samples/debug/memfault/README.rst:
```diff
@@ -65,20 +65,21 @@ See `Memfault: Error Tracking with Trace Events`_ for information on how to conf
 
-Coredumps can be triggered either by using the Memfault shell command ``mflt crash``, or by pressing a button:
+The sample enables Memfault shell by default.
+The shell offers multiple commands to test a wide range of functionality offered by Memfault SDK.
```
```suggestion
The shell offers multiple commands to test a wide range of functionality offered by the Memfault SDK.
```

peknis commented on samples/debug/memfault/README.rst:
```diff
@@ -92,6 +93,7 @@ Minimal setup
 =============
 
 To send data to the Memfault cloud, a project key must be configured using :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY`.
+If nRF7002 DK is used and one wishes to automatically connect to a Wi-Fi network upon boot, the Wi-Fi credentials be configured using :kconfig:option`CONFIG_WIFI_CREDENTIALS_STATIC_SSID` and :kconfig:option:`CONFIG_WIFI_CREDENTIALS_STATIC_PASSWORD`.
```
```suggestion
If you are using the nRF7002 DK and want to automatically connect to a Wi-Fi network upon boot, configure the Wi-Fi credentials using the Kconfig options :kconfig:option`CONFIG_WIFI_CREDENTIALS_STATIC_SSID` and :kconfig:option:`CONFIG_WIFI_CREDENTIALS_STATIC_PASSWORD`.
```

peknis commented on samples/debug/memfault/README.rst:
```diff
@@ -165,17 +167,23 @@ Before testing, ensure that your device is configured with the project key of yo
+        <inf> memfault_sample: Connecting to network
 
-#. When the connection is established, the sample displays the captured LTE time-to-connect metric (``Ncs_LteTimeToConnect``) on the terminal:
+#. When the connection is established, the following messag eis printed:
```
```suggestion
#. When the connection is established, the following message is shown:
```

peknis commented on samples/debug/memfault/src/network_common.h:
```diff
@@ -0,0 +1,23 @@
+void network_connected(void);
+
+/**
+ * @brief To be called when network connection has been terminated.
```
```suggestion
 * @brief A function to be called when the network connection has been terminated.
```

jtguggedal commented:
@peknis I have now updated with your suggestions. I have also added a separate commit moving the debug samples docs into a separate section and moving the Memfault library docs into the debug libraries section.

peknis commented:
> @peknis I have now updated with your suggestions. I have also added a separate commit moving the debug samples docs into a separate section and moving the Memfault library docs into the debug libraries section.

Maybe we should also mention the moving of the Memfault library to another location in the doc structure.

jtguggedal commented:
> 

That makes sense, will add that
